File: src/tools/websearch.py
import io
import json
import sys

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8")

from ddgs import DDGS

def search(query: str, max_results: int = 8) -> list[dict]:
    results = []
    try:
        with DDGS() as ddgs:
            for r in ddgs.text(query, max_results=max_results):
                results.append({
                    "title": r.get("title", ""),
                    "url": r.get("href", ""),
                    "snippet": r.get("body", ""),
                })
    except Exception as e:
        return [{"error": str(e)}]
    return results


def main():
    args = sys.argv[1:]
    query = " ".join(args) if args else sys.stdin.read().strip()
    if not query:
        print(json.dumps({"error": "No query provided", "results": []}))
        sys.exit(1)
    results = search(query)
    out = {"query": query, "results": results, "count": len(results)}
    print(json.dumps(out, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: .opencode/plugins/websearch.js
import { tool } from "@opencode-ai/plugin";

export const WebSearchPlugin = async ({ directory }) => {
  return {
    tool: {
      websearch: tool({
        description: "Search the web using DuckDuckGo (free, no API key). Returns results with titles, URLs, and snippets.",
        args: {
          query: tool.schema.string().describe("Search query"),
          max_results: tool.schema.number().optional().describe("Max results (default 8)"),
        },
        async execute(args, context) {
          const scriptPath = `${context.directory}/src/tools/websearch.py`;
          const proc = Bun.spawn(["python", scriptPath, args.query], {
            env: { ...process.env },
          });
          const out = await new Response(proc.stdout).text();
          const err = await new Response(proc.stderr).text();
          const exitCode = await proc.exited;
          if (exitCode !== 0) {
            return `Search failed (exit ${exitCode}): ${err || out}`;
          }
          try {
            const data = JSON.parse(out);
            if (data.error) return `Search error: ${data.error}`;
            if (!data.results || data.results.length === 0) return "No results found.";
            return data.results.map((r, i) =>
              `[${i + 1}] ${r.title}\n    URL: ${r.url}\n    ${r.snippet}`
            ).join("\n\n");
          } catch (e) {
            return `Failed to parse results: ${e.message}\nRaw: ${out.slice(0, 500)}`;
          }
        },
      }),
    },
  };
};
